### PR TITLE
[dev-launcher] add gradle plugin to log all OkHttp requests

### PR DESCRIPTION
**Pull request by Kudo**
# Why

poc of logging requests from all `OkHttpClient`
close ENG-7465

# How

introducing an expo-dev-launcher-gradle-plugin to do AOP (by ASM API) for `OkHttpClient.Builder.build()`. basically transforming the bytecode to add our `DevLauncherOkHttpInterceptor` before the actual build().

references for writing gradle plugin to do bytecode manipulation
- https://speakerdeck.com/rom4ek/bytecode-transformations-exploring-new-agp-apis-for-instrumenting-your-app
- https://blog.sentry.io/2021/12/14/bytecode-transformations-the-android-gradle-plugin/

# Test Plan

- send fetch calls and check chrome devtools
- ci passed
- decompile apk and check whether the java code looks correct
Screenshot 2023-02-22 at 1 51 10 PM

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
  - will update the changelog together with #21265 
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/network/DevLauncherNetworkLogger.kt [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ * OkHttp `Headers` extension method to generate a simple key-value map
+ * which only exposing single value for a key.
+ */
+fun Headers.toSingleMap(): Map<String, String> {
+  val result = ArrayMap<String, String>()
+  for (key in names()) {
+    result[key] = get(key)
+  }
+  return result
+}
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/network/DevLauncherNetworkLogger.kt:176]:
```suggestion
}

```

File: packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/network/DevLauncherOkHttpInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    if (DevLauncherNetworkLogger.instance.shouldEmitEvents()) {
+      val requestId = request.hashCode().toString()
+      DevLauncherNetworkLogger.instance.emitNetworkWillBeSent(request, requestId)
+      val response = chain.proceed(request)
+      DevLauncherNetworkLogger.instance.emitNetworkResponse(request, requestId, response)
+      return response
+    }
+    return chain.proceed(request)
+  }
+}
```
Comment (lukmccall) [inline packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/network/DevLauncherOkHttpInterceptor.kt:19]:
```suggestion
}

```

**Review by lukmccall — APPROVED** [review]
It's a little bit unsafe in my opinion, but as long as people can turn that off I'm okay with that.

**Comment by Kudo** [comment]
> It's a little bit unsafe in my opinion, but as long as people can turn that off I'm okay with that.

we only target debug build now and specific `OkHttpClient.Builder.build()` method. it should be safe IMO. the feature is disabled by default. it should be enabled through the `EX_DEV_CLIENT_NETWORK_INSPECTOR=true` gradle property. i'm merging this and see how people feel about the network inspecting feature.

**Pull request merged**